### PR TITLE
Add prometheus metrics server

### LINKED ISSUE #892: Add monitoring support with prometheus
As discussed in #736, we need to expose some metrics about the Runtime for monitoring purposes. [Prometheus](https://prometheus.io) provides an API for collecting metrics. To be able to use its [pull model](https://prometheus.io/docs/introduction/faq/#why-do-you-pull-rather-than-push), we also need an HTTP server to access the metrics.

### PR DESCRIPTION
**Pull request by rbehjati**
Add prometheus metrics server:
* Adds the `metrics` module in `oak_runtime` that uses `prometheus`
* Starts the metrics server in the `oak_loader`
* Updates `reproducibility_index` with new hash for the `oak_loader`
* Adds integration tests for the `metrics` module
* Adds basic instrumentation for `grpc_server`, but not tested

Tested with `./scripts/docker_run cargo run --package=oak_loader -- --application=/opt/my-project/bazel-client-bin/examples/chat/config/config.bin --metrics-port=9876`. This starts the metrics server on port `9876`.

Fixes #892

# Checklist

- [x] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [x] I have written tests that cover the code changes.
  - [x] I have checked that these tests are run by [Cloudbuild](cloudbuild.yaml)
  - [ ] I have updated [documentation](docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [ ] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

File: oak/server/rust/oak_loader/src/main.rs [hunk 1]
```diff
@@ -51,7 +51,8 @@ fn read_file(filename: &str) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
     Ok(buffer)
 }
 
-fn main() -> Result<(), Box<dyn std::error::Error>> {
+#[tokio::main]
```
Comment (ipetr0v) [inline oak/server/rust/oak_loader/src/main.rs:54]: Why do we need this right now?
The only this that needs `tokio` is a gRPC pseudo-node.
But it uses an instance of `tokio::runtime`:
https://github.com/project-oak/oak/blob/7e9f6cd424ab7729c4e12718ee1e8703e668eac5/oak/server/rust/oak_runtime/src/node/grpc_server.rs#L316-L319

File: oak/server/rust/oak_runtime/src/node/grpc_server.rs [hunk 2]
```diff
@@ -208,6 +213,8 @@ impl GrpcServerNode {
     /// channel to the [`GrpcServerNode::channel_writer`].
     /// Returns a [`Handle`] for reading a gRPC response from.
     fn process_request(&self, request: GrpcRequest) -> Result<Handle, GrpcServerError> {
+        HTTP_COUNTER.inc();
```
Comment (ipetr0v) [inline oak/server/rust/oak_runtime/src/node/grpc_server.rs:216]: Is it a Singleton counter?

Comment (rbehjati) [reply oak/server/rust/oak_runtime/src/node/grpc_server.rs:216]: This is a static variable. So we I guess we could have other counters too. I don't know what is the best practice for using these metrics. I'll have to investigate that part.

File: oak/server/rust/oak_loader/src/main.rs (hunk 1 above)
Comment (ipetr0v) [reply oak/server/rust/oak_loader/src/main.rs:54]: I don't know if runtime that is used by a gRPC node will work, if everything is already being run by a Tokio runtime.
But if it works, then don't see any problems with using `#[tokio::main]`

File: oak/server/rust/oak_loader/src/main.rs [hunk 3]
```diff
@@ -64,8 +65,14 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
...
     configure_and_run(app_config).map_err(|error| format!("Runtime error: {:?}", error))?;
-    // Park current thread.
-    park();
+
+    // TODO: port should come from config
+    let port = 3030;
+    let metrics_future = metrics::serve_metrics(port);
+
+    if let Err(e) = metrics_future.await {
+        eprintln!("Error when waiting for metrics_future: {:?}", e);
```
Comment (ipetr0v) [inline oak/server/rust/oak_loader/src/main.rs:74]: Should it use `error!()`?

File: oak/server/rust/oak_loader/src/main.rs [hunk 4]
```diff
@@ -64,8 +65,14 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
 
     // Spawns a new thread corresponding to an initial Wasm Oak node.
     configure_and_run(app_config).map_err(|error| format!("Runtime error: {:?}", error))?;
-    // Park current thread.
-    park();
+
+    // TODO: port should come from config
+    let port = 3030;
+    let metrics_future = metrics::serve_metrics(port);
```
Comment (ipetr0v) [inline oak/server/rust/oak_loader/src/main.rs:71]: Is it running in the main thread?
Should we create a new thread for this?

File: oak/server/rust/oak_runtime/src/metrics.rs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+use log::info;
+use prometheus::{
+    labels, opts, register_counter, register_gauge, register_histogram_vec, Counter, Encoder,
+    Gauge, HistogramVec, TextEncoder,
+};
+use std::{convert::Infallible, net::SocketAddr};
+
+lazy_static::lazy_static! {
+pub static ref HTTP_COUNTER: Counter = register_counter!(opts!(
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/metrics.rs:15]: I think this should be an Int counter?

File: oak/server/rust/oak_runtime/src/metrics.rs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    labels! {"handler" => "all",}
+))
+.unwrap();
+pub static ref HTTP_REQ_HISTOGRAM: HistogramVec = register_histogram_vec!(
+    "http_request_duration_seconds",
+    "The HTTP request latencies in seconds.",
+    &["handler"]
+)
+.unwrap();
+pub static ref NUM_NODES: Gauge = register_gauge!(opts!(
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/metrics.rs:33]: And this should be an int gauge?

File: oak/server/rust/oak_runtime/src/metrics.rs [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    labels, opts, register_counter, register_gauge, register_histogram_vec, Counter, Encoder,
+    Gauge, HistogramVec, TextEncoder,
+};
+use std::{convert::Infallible, net::SocketAddr};
+
+lazy_static::lazy_static! {
+pub static ref HTTP_COUNTER: Counter = register_counter!(opts!(
+    "http_requests_total",
+    "Total number of HTTP requests made.",
+    labels! {"handler" => "all",}
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/metrics.rs:18]: Not sure I get what these labels are for, maybe add a comment?

File: oak/server/rust/oak_runtime/src/metrics.rs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+))
+.unwrap();
+}
+
+async fn process_metrics(_req: Request<Body>) -> Result<Response<Body>, hyper::Error> {
+    let encoder = TextEncoder::new();
+    let metric_families = prometheus::gather();
+    let mut buffer = vec![];
+    encoder.encode(&metric_families, &mut buffer).unwrap();
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/metrics.rs:45]: Use `expect` and some relevant error message, if possible.

File: oak/server/rust/oak_runtime/src/metrics.rs [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+};
+
+use log::info;
+use prometheus::{
+    labels, opts, register_counter, register_gauge, register_histogram_vec, Counter, Encoder,
+    Gauge, HistogramVec, TextEncoder,
+};
+use std::{convert::Infallible, net::SocketAddr};
+
+lazy_static::lazy_static! {
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/metrics.rs:14]: Another approach would be to declare these counters as part of the `Runtime` itself (well,in a separate sub-object really) instead of global mutable variables. Feel free to go ahead with this if it makes things simpler, but worth keeping in mind how we expect things to look eventually.

File: oak/server/rust/oak_runtime/src/runtime/mod.rs [hunk 10]
```diff
@@ -773,6 +774,12 @@ impl Runtime {
             .write()
             .expect("could not acquire lock on node_infos")
             .insert(node_id, node_info);
+        crate::metrics::NUM_NODES.set(
+            self.node_infos
+                .read()
+                .expect("could not acquire lock on node_infos")
+                .len() as f64,
+        );
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/runtime/mod.rs:782]: Should there be a corresponding update when removing nodes too?

File: oak/server/rust/oak_runtime/tests/integration_test.rs [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+                    Err(e) => Err(e),
+                    Ok(buf) => Ok(format!("{:?}", buf)),
+                }
+            }
+        }
+    }
+}
+
+#[test]
+fn test_metrics_gives_the_correct_number_of_nodes() {
```
Comment (rbehjati) [inline oak/server/rust/oak_runtime/tests/integration_test.rs:91]: @daviddrysdale Is this anything like the e2e test you described the other day?

File: oak/server/rust/oak_runtime/src/metrics.rs [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    header::CONTENT_TYPE,
+    service::{make_service_fn, service_fn},
+    Body, Request, Response, Server,
+};
+
+use log::{error, info};
+use prometheus::{Encoder, TextEncoder};
+use std::{convert::Infallible, net::SocketAddr};
+
+async fn process_metrics(_req: Request<Body>) -> Result<Response<Body>, hyper::Error> {
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/metrics.rs:27]: Maybe call this method `serve_metrics`?

File: oak/server/rust/oak_runtime/src/node/grpc_server.rs [hunk 13]
```diff
@@ -29,6 +29,40 @@ use oak_abi::{
 
 use crate::{pretty_name_for_thread, runtime::RuntimeProxy, Handle};
 
+use prometheus::{opts, register_histogram, register_int_counter, Histogram, IntCounter};
+
+struct GrpcMetrics {
+    request_duration: Histogram,
+    requests_total: IntCounter,
+    response_size: Histogram,
+}
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/node/grpc_server.rs:38]: I think it would be more manageable to centralize the metrics definitions in a single object the metrics package, rather than spread it out. Also it will help to move away from the static initializer eventually. In fact, at that point, you may as well use a non-static Registry, but up to you if you want to leave that for another PR.

File: oak/server/rust/oak_runtime/src/node/grpc_server.rs [hunk 14]
```diff
@@ -167,16 +201,22 @@ impl GrpcServerNode {
...
             .and_then(|request| self.process_request(request))
             // Read a gRPC response from the Runtime.
             .and_then(|response_reader| self.process_response(response_reader))
             // Send gRPC response back to the HTTP client.
             .map(|body| Self::http_response(http::StatusCode::OK, body))
             // Convert an error to an HTTP response with a corresponding error status.
-            .or_else(|error| Ok(Self::http_response(error.into(), vec![])))
+            .or_else(|error| Ok(Self::http_response(error.into(), vec![])));
+
+        timer.observe_duration();
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/node/grpc_server.rs:217]: Is this actually measuring any elapsed time? AFAICT everything is async here, so it's just measuring the time needed to create the `Future`. Or am I missing some magic happening behind the scenes in `async fn`?

File: oak/server/rust/oak_runtime/src/metrics.rs [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+
+    let server = Server::bind(&addr).serve(make_svc);
+
+    info!("Started metrics server on port {}", port);
+
+    // Run this server for... forever!
+    if let Err(e) = server.await {
+        error!("server error: {}", e);
+    }
+    Ok(())
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/metrics.rs:62]: Could you just return `server` and assume any error is logged upstreams anyways?

File: oak/server/rust/oak_runtime/tests/integration_test.rs [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+
+    use hyper::{Client, Uri};
+    use log::info;
+    use maplit::hashmap;
+    use oak_abi::OakStatus;
+    use oak_runtime::{config, metrics};
+    use std::sync::Arc;
+    use wat::parse_str;
+
+    const PORT: u16 = 9876;
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/tests/integration_test.rs:30]: Could you add a doc on this?

File: oak/server/rust/oak_runtime/tests/integration_test.rs [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+        let uri = format!("http://localhost:{}", &PORT)
+            .parse::<Uri>()
+            .expect("Could not parse URI.");
+        match client.get(uri).await {
+            Err(e) => Err(e),
+            Ok(res) => {
+                info!("status: {}", res.status());
+
+                // Concatenate the body stream into a single buffer...
+                match hyper::body::to_bytes(res).await {
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/tests/integration_test.rs:81]: Does `.await?` work? If so, I think you can remove two levels of nesting.

File: oak/server/rust/oak_runtime/tests/integration_test.rs [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+                    Ok(buf) => Ok(format!("{:?}", buf)),
+                }
+            }
+        }
+    }
+}
+
+#[test]
+fn test_metrics_gives_the_correct_number_of_nodes() {
+    simple_logger::init().unwrap();
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/tests/integration_test.rs:92]: Is there any advantage in using `simple_logger` vs `env_logger`? I am not familiar with the former.

File: oak/server/rust/oak_runtime/tests/integration_test.rs [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+
+    // Alternatively, one could start the runtime without spawning a new thread, but this seems to
+    // be a more generic way of starting the runtime.
+    let rt_handle = spawn(move || {
+        let _res = common::start_runtime();
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/tests/integration_test.rs:97]: Maybe `.expect` this?

File: oak/server/rust/oak_runtime/tests/integration_test.rs [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+    });
+
+    let mut rt = tokio::runtime::Runtime::new().expect("Couldn't create Tokio runtime");
+    let res = rt
+        .block_on(common::read_metrics())
+        .expect("Could not get the data from future.");
+    assert_eq!(res, "b\"# HELP nodes_count Number of nodes in the runtime.\\n# TYPE nodes_count gauge\\nnodes_count 1\\n\"");
+
+    metrics_handle.thread().unpark();
+    let _ = rt_handle.join();
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/tests/integration_test.rs:115]: `.expect`?

File: oak/server/rust/oak_runtime/tests/integration_test.rs [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+fn test_metrics_gives_the_correct_number_of_nodes() {
+    simple_logger::init().unwrap();
+
+    // Alternatively, one could start the runtime without spawning a new thread, but this seems to
+    // be a more generic way of starting the runtime.
+    let rt_handle = spawn(move || {
+        let _res = common::start_runtime();
+    });
+
+    // start metrics server in a new thread
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/tests/integration_test.rs:100]: (nit)

```suggestion
    // Start metrics server in a new thread.
```

File: oak/server/rust/oak_runtime/tests/integration_test.rs [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+        common::start_metrics_server();
+
+        // Keep the server up until the test is complete.
+        park();
+    });
+
+    let mut rt = tokio::runtime::Runtime::new().expect("Couldn't create Tokio runtime");
+    let res = rt
+        .block_on(common::read_metrics())
+        .expect("Could not get the data from future.");
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/tests/integration_test.rs:111]: This message seems quite vague :) Also if there is an error is probably about the logic, not about the future itself?

File: oak/server/rust/oak_runtime/tests/integration_test.rs [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+    assert_eq!(res, "b\"# HELP nodes_count Number of nodes in the runtime.\\n# TYPE nodes_count gauge\\nnodes_count 1\\n\"");
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/tests/integration_test.rs:112]: Neat!

File: oak/server/rust/oak_runtime/tests/integration_test.rs [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+    let rt_handle = spawn(move || {
+        let _res = common::start_runtime();
+    });
+
+    // start metrics server in a new thread
+    let metrics_handle = spawn(move || {
+        common::start_metrics_server();
+
+        // Keep the server up until the test is complete.
+        park();
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/tests/integration_test.rs:105]: I'm not convinced this is doing anything at all. Doesn't `common::start_metrics_server()` block forever anyways?

File: oak/server/rust/oak_runtime/tests/integration_test.rs (hunk 18 above)
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/tests/integration_test.rs:92]: Hmm, we seem to use both in different parts of the codebase:
 - `simple_logger`: used by `oak_glue`, `running_average`, `hello_world`, `private_set_intersection`, `aggregator/module`, `translator`, `abitest`
 - `env_logger`: used by `oak_loader` ,`aggregator/backend`, `split_grpc/{client,proxy}`

Maybe we should settle on one consistently?

File: oak/server/rust/oak_runtime/tests/integration_test.rs (hunk 11 above)
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/tests/integration_test.rs:91]: Yep, definitely heading that way &ndash; nice!  

However, these kind of tests do end up being quite brittle &ndash; small changes to the code give test breakages that aren't obviously correlated with the change.  

It would be nice to reduce the brittleness as much as possible, to make life easier for developers. This initial test should be fairly robust, because there's not much that can change with a single Node that immediately exits.  However, checking the exact help text makes the test a bit of a change-detector &ndash; it's probably worth adding a helper that extracts a particular numeric value from the result of `read_metrics`, so that (say) changes to the help text for a metric don't result in lots of broken tests.

Maybe a `read_int_metric(metric_name: &str) -> Option<i64>` method that just pulls a specific value?  (I'd hope that there's a Prometheus client library somewhere that can help with this).

Longer term, it might also be nice to have a test that's even more end-to-end: we could (say) change the `abitest` client binary to query the metrics (externally, over HTTP) at start-of-day, and check that the expected Node count is there (4 at the moment).  That definitely would be a brittle test though &ndash; if you ran the client twice in a row without restarting the Runtime, the count would be out &ndash; so we'd probably need to think about something that's more robust.

File: oak/server/rust/oak_loader/src/main.rs [hunk 25]
```diff
@@ -63,9 +68,45 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+
     // Park current thread.
     park();
 
     Ok(())
 }
+
+fn start_metrics() {
+    let opt = Opt::from_args();
+    let port = opt.metrics_port.unwrap_or(3030);
```
Comment (ipetr0v) [inline oak/server/rust/oak_loader/src/main.rs:83]: Does metrics always start or it's possible to disable it?

File: oak/server/rust/oak_loader/src/main.rs [hunk 26]
```diff
@@ -40,6 +44,8 @@ pub struct Opt {
     private_key: Option<String>,
     #[structopt(long, help = "Path to the PEM-encoded certificate chain")]
     cert_chain: Option<String>,
+    #[structopt(long, help = "Metrics server port number")]
```
Comment (ipetr0v) [inline oak/server/rust/oak_loader/src/main.rs:47]: Could you please add a description that the default port is `3030`?
And also use `#[structopt(default_value = 3030)]`

File: oak/server/rust/oak_loader/src/main.rs [hunk 27]
```diff
@@ -63,9 +68,45 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+            error!("Error in runtime: {}", e);
+        }
+    }));
+
+    // start metrics server in a new thread
+    handles.push(spawn(move || {
+        start_metrics();
+    }));
+
+    for handle in handles {
```
Comment (ipetr0v) [inline oak/server/rust/oak_loader/src/main.rs:107]: There are only two handles, should they just be two distinct variables for clarity?

File: oak/server/rust/oak_runtime/Cargo.toml [hunk 28]
```diff
@@ -26,12 +26,17 @@ oak_abi = "=0.1.0"
...
 wasmi = { version = "*", default-features = false, features = ["core"] }
+prometheus = { version = "0.8", features = [
+  "nightly"
+], default-features = false }
+futures = "*"
 
 [dev-dependencies]
 maplit = "*"
 wat = "*"
+simple_logger = "*"
```
Comment (ipetr0v) [inline oak/server/rust/oak_runtime/Cargo.toml:39]: I don't remember if `scripts/format` should sort dependencies in the alphabetical order

File: oak/server/rust/oak_runtime/src/metrics.rs [hunk 29]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+pub async fn serve_metrics(port: u16) -> Result<(), Box<dyn std::error::Error>> {
+    let addr = SocketAddr::from(([127, 0, 0, 1], port));
+
+    // A `Service` is needed for every connection, so this
+    // creates one from the `process_metrics` function.
+    let make_svc = make_service_fn(|_conn| async {
+        // service_fn converts our function into a `Service`
+        Ok::<_, Infallible>(service_fn(process_metrics))
```
Comment (ipetr0v) [inline oak/server/rust/oak_runtime/src/metrics.rs:51]: Is it `Infallible`? It returns a `Result<..., hyper::Error>`

File: oak/server/rust/oak_runtime/src/metrics.rs [hunk 30]
```diff
@@ -0,0 +1,63 @@
...
+        .status(200)
+        .header(CONTENT_TYPE, encoder.format_type())
+        .body(Body::from(buffer))
+        .expect("Error when building the HTTP response.");
+
+    Ok(response)
+}
+
+pub async fn serve_metrics(port: u16) -> Result<(), Box<dyn std::error::Error>> {
+    let addr = SocketAddr::from(([127, 0, 0, 1], port));
```
Comment (ipetr0v) [inline oak/server/rust/oak_runtime/src/metrics.rs:45]: Should we listen on any address: `SocketAddr::new("0.0.0.0".parse(), port)`?

File: oak/server/rust/oak_runtime/src/metrics.rs [hunk 31]
```diff
@@ -0,0 +1,63 @@
...
+use prometheus::{Encoder, TextEncoder};
+use std::{convert::Infallible, net::SocketAddr};
+
+async fn process_metrics(_req: Request<Body>) -> Result<Response<Body>, hyper::Error> {
+    let encoder = TextEncoder::new();
+    let metric_families = prometheus::gather();
+    let mut buffer = vec![];
+    encoder
+        .encode(&metric_families, &mut buffer)
+        .expect("Could not encode metrics data!");
```
Comment (ipetr0v) [inline oak/server/rust/oak_runtime/src/metrics.rs:33]: The function already returns `hyper::Error`, do we need to return this error, or it's a critical error that stops Oak?

File: oak/server/rust/oak_runtime/src/node/grpc_server.rs [hunk 32]
```diff
@@ -167,16 +201,22 @@ impl GrpcServerNode {
                 error
             })?;
 
+        METRICS.requests_total.inc();
```
Comment (ipetr0v) [inline oak/server/rust/oak_runtime/src/node/grpc_server.rs:204]: Do we also count failed requests?

File: oak/server/rust/oak_loader/src/main.rs (hunk 27 above)
Comment (rbehjati) [reply oak/server/rust/oak_loader/src/main.rs]: I prefer this one.

File: reproducibility_index [hunk 33]
```diff
@@ -1 +1 @@
-097de181dd94f6c91b9abfbd78d64ea98acf653c4881db98069fbab764ebaf4b  ./target/x86_64-unknown-linux-musl/release/oak_loader
+1ba4ea8d7bbf5f6dd25a2ab52e3ac0426d5dcd4803b96dee706cbf9dceeea770  ./target/x86_64-unknown-linux-musl/release/oak_loader
```
Comment (tiziano88) [inline reproducibility_index:1]: This is not necessary any more, please rebase on master.

File: oak/server/rust/oak_runtime/tests/integration_test.rs [hunk 34]
```diff
@@ -0,0 +1,100 @@
...
+        Ok(std::str::from_utf8(&buf[..]).unwrap().to_string())
+    }
+}
+
+fn get_int_metric_value(all_metrics: &str, metric_name: &str) -> Option<i64> {
+    let pattern = format!(r"(.*)(\b{} \b)([0-9]+)(.*)", metric_name);
+    let re = regex::Regex::new(&pattern).unwrap();
+    re.captures(&all_metrics)
+        .map(|c| c[3].parse::<i64>().unwrap())
+}
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/tests/integration_test.rs:79]: I'm not sure trying to parse metrics with regexps is particularly readable / maintainable. I would suggest, either stick to making expectations on the text itself (which I think is what you had before), or use an off-the-shelf prometheus parsing library (though that seems premature now).

Also see http://go/tott/510 .

Comment (rbehjati) [reply oak/server/rust/oak_runtime/tests/integration_test.rs:79]: After the last changes the response looks like this: 
```
# HELP grpc_request_duration_seconds The gRPC request latencies in seconds.
# TYPE grpc_request_duration_seconds histogram
grpc_request_duration_seconds_bucket{le="0.005"} 0
grpc_request_duration_seconds_bucket{le="0.01"} 0
grpc_request_duration_seconds_bucket{le="0.025"} 0
grpc_request_duration_seconds_bucket{le="0.05"} 0
grpc_request_duration_seconds_bucket{le="0.1"} 0
grpc_request_duration_seconds_bucket{le="0.25"} 0
grpc_request_duration_seconds_bucket{le="0.5"} 0
grpc_request_duration_seconds_bucket{le="1"} 0
grpc_request_duration_seconds_bucket{le="2.5"} 0
grpc_request_duration_seconds_bucket{le="5"} 0
grpc_request_duration_seconds_bucket{le="10"} 0
grpc_request_duration_seconds_bucket{le="+Inf"} 0
grpc_request_duration_seconds_sum 0
grpc_request_duration_seconds_count 0
# HELP grpc_requests_total Total number of gRPC requests received.
# TYPE grpc_requests_total counter
grpc_requests_total 0
# HELP grpc_response_size_bytes The gRPC response sizes in bytes.
# TYPE grpc_response_size_bytes histogram
grpc_response_size_bytes_bucket{le="0.005"} 0
grpc_response_size_bytes_bucket{le="0.01"} 0
grpc_response_size_bytes_bucket{le="0.025"} 0
grpc_response_size_bytes_bucket{le="0.05"} 0
grpc_response_size_bytes_bucket{le="0.1"} 0
grpc_response_size_bytes_bucket{le="0.25"} 0
grpc_response_size_bytes_bucket{le="0.5"} 0
grpc_response_size_bytes_bucket{le="1"} 0
grpc_response_size_bytes_bucket{le="2.5"} 0
grpc_response_size_bytes_bucket{le="5"} 0
grpc_response_size_bytes_bucket{le="10"} 0
grpc_response_size_bytes_bucket{le="+Inf"} 0
grpc_response_size_bytes_sum 0
grpc_response_size_bytes_count 0
# HELP runtime_nodes_count Number of nodes in the runtime.
# TYPE runtime_nodes_count gauge
runtime_nodes_count 1
```
It does not make sense to assert on the whole thing. Also, as David pointed out in his comment, that kind of test would be brittle. The only off-the-shelf component I found is [prometheus-scrape](https://crates.io/crates/prometheus-scrape). But that basically does not have any documentation. So, I don't want to use it. 
For now, I'll keep this test as is.

File: oak/server/rust/oak_loader/src/main.rs [hunk 35]
```diff
@@ -40,6 +44,14 @@ pub struct Opt {
...
     #[structopt(long, help = "Path to the PEM-encoded certificate chain")]
     cert_chain: Option<String>,
+    #[structopt(
+        long,
+        default_value = "3030",
+        help = "Metrics server port number. Defaults to 3030."
+    )]
+    metrics_port: u16,
+    #[structopt(long, help = "Starts the Runtime without a metrics server.")]
+    no_metrics: bool,
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply oak/server/rust/oak_loader/src/main.rs:54]: Sure, but why have a negative meaning for a bool flag? It may as well be positive, and have a default of true. then it can be specified as `--metrics=false` if someone wants to disable the server.

Also see http://go/tott/530 .

File: oak/server/rust/oak_runtime/src/metrics/server.rs [hunk 36]
```diff
@@ -0,0 +1,88 @@
...
+    })?;
+
+    info!("Metrics size: {}", buffer.len());
+
+    Response::builder()
+        .status(http::StatusCode::OK)
+        .header(CONTENT_TYPE, encoder.format_type())
+        .body(Body::from(buffer))
+        .map_err(|e| {
+            MetricsServerError::ResponseError(format!("Could not build the response: {}", e))
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/metrics/server.rs:59]: Does this end up being sent to the client? If so, we should not propagate details about internal errors, in general they may leak data. Better to log the error, and return safe, pre-defined error messages to the client. But maybe that's already the case, I don't know whether this `map_err` affects the response or not.

File: oak/server/rust/oak_loader/src/main.rs (hunk 35 above)
[the remark this answers is not in the document]
Comment (rbehjati) [reply oak/server/rust/oak_loader/src/main.rs:54]: That makes sense to me if you want to disable the metrics server by default (no `--metrics` option). Then to enable the metrics server one has to use `--metrics` or `--metrics=true` (although the latter looks redundant to me). 
If the server should be disabled by default, then something like `enable-metrics` or `with-metrics` reads better in my opinion.

File: oak/server/rust/oak_runtime/src/metrics/server.rs (hunk 36 above)
Comment (rbehjati) [reply oak/server/rust/oak_runtime/src/metrics/server.rs:59]: I don't think so. According to the [documentation](https://doc.rust-lang.org/std/result/enum.Result.html#method.map_err), `map_err` leave the `Ok` value untouched.
